### ISSUE TITLE
Warning: react-modal: App element is not defined.

### ISSUE THREAD

**Issue by StokeMasterJack** [reactjs/react-modal#576]
I upgraded my yarn version and re-ran yarn install. And now I am getting this warning. 

It's strange, because nothing in my code has changed. Something in node_modules is obviously different but who knows what. I can't do a diff on node_modules because that was never check in to git.

**Comment by diasbruno** [comment]
@StokeMasterJack we made some changes recently to api to define the `app element`. Can you get the log message you got?

**Comment by diasbruno** [comment]
Can you check if setting `Modal.setAppElement` the error goes away? Ref: [https://github.com/reactjs/react-modal#app-element](https://github.com/reactjs/react-modal#app-element)

**Comment by StokeMasterJack** [comment]
Here is the exact log message:

Warning: react-modal: App element is not defined. Please use `Modal.setAppElement(el)` or set `appElement={el}`. This is needed so screen readers don't see main content when modal is opened. It is not recommended, but you can opt-out by setting `ariaHideApp={false}`.

**Comment by diasbruno** [comment]
Oh sorry, though it was an error. This is related to this change #570. This is a reminder that not setting the app element will/can cause accessibility issue.

**Comment by diasbruno** [comment]
So, you need to explicit tell the modal that you don't want to use the app element with `ariaHideApp={false}` (if that's your case).

**Comment by diasbruno** [comment]
I'll keep this open. Documentation is missing for this behavior.

**Comment by dlong500** [comment]
Can this feature be explained a bit more?  When I try to use it I'm also getting an error like so:
`ReferenceError: document is not defined` coming from `ariaAppHider.js`

Also, is there a reason this configuration can't be encapsulated into the Modal props and must use `Modal.setAppElement`?  How are you supposed to handle the feature given multiple Modals throughout a large app in various components?

**Comment by dlong500** [comment]
Ah, I think the issue was my server rendering of the app.  Once I wrapped the `setAppElement` call in a check to make sure it was in a client/browser environment then it appears to run OK.

I'm a still a bit unsure if things are working properly though especially because the docs state "If you are doing server-side rendering, you should use this property".  That doesn't seem to make sense to me because the DOM is not even available at that point yet.

**Comment by diasbruno** [comment]
@dlong500 The documentation related to this is outdated. Also, we need to tests how the modal will work on a SSR environment. I did some tests to get it working, but very superficial - just to make it work. 

For the multiple modal case, you can use `appElement` attribute.

**Comment by mandazi** [comment]
I am also getting this in the console log. No code changes in terms of my use of react-modal either. Not sure why I am getting it.
`Warning: react-modal: App element is not defined. Please use `Modal.setAppElement(el)` or set `appElement={el}`. This is needed so screen readers don't see main content when modal is opened. It is not recommended, but you can opt-out by setting `ariaHideApp={false}`.`

**Comment by mandazi** [comment]
Nevermind I just read the thread again. Looks like I need the warning is correct. We need just need to specify it.

**Comment by amangeot** [comment]
Hello, I got the same warning, how do you guys make use of appElement?

Is it `appElement={document.getElementById('root')}` where 'root' would be where the app is mounted?

**Comment by mockdeep** [comment]
I was seeing this when the modal is unmounted without having been triggered. Adding this fixed the issue:

```js
Modal.setAppElement('#app-base');
```

**Comment by sksundram** [comment]
@mockdeep Where did you add `Modal.setAppElement('#app-base');`

**Comment by mockdeep** [comment]
@sksundram right now I've got it in each of the files that I'm importing `Modal` for rendering. I suspect you could put it just about anywhere, though, like in your top-level `app.js` file, as long as it is called before the modal is rendered.

```jsx
import Modal from 'react-modal';
import React from 'react';

function HelpModal({isOpen, closeModal}) {
  return (
    <Modal isOpen={isOpen} onRequestClose={closeModal}>
      'modal content'
    </Modal>
  );
}

Modal.setAppElement('#app-base');

export default HelpModal;
```

**Comment by lbelinsk** [comment]
@diasbruno I wouldn't mind preparing the documentation for the ssr and appElement subjects.
If that sounds ok, If there is any specific info you want me to elaborate on while I am at it, please let me know.

For the meantime, just for reference, for the projects which use react-modal, there is no need anymore to add the safety check that @NathanielHill suggested.

Instead of doing:
```
if (typeof(window) !== 'undefined') {
   ReactModal.setAppElement('body')
}
```

It is safe enough to do:
```
ReactModal.setAppElement('body')
```

**Comment by diasbruno** [comment]
@lbelinsk Sure, it would be really great.

**Comment by NathanielHill** [comment]
@catamphetamine 

Fair point. For me the issue was just setting it to begin with. If what your saying is correct, that does sound like an accessibility issue, so if your React app has a container `<div>` as a child of `<body>` then that would be a better target.

**Comment by ReneCode** [comment]
using the react-testing-library (https://testing-library.com/) I get rid of that warning with:
    
    import Modal from "react-modal";
    
    const { container } = render(<MyComponent />);
    Modal.setAppElement(container);
    
    ....  // to the testing, use Modal

**Comment by emmy29** [comment]
Adding  ariaHideApp={false} to modal works for me.

 <Modal isOpen={this.state.modalOpen} ariaHideApp={false}>
          <LoginModal closeModal={() => this.openModal()}></LoginModal>
        </Modal>

**Comment by raphab3** [comment]
> ```js
> Modal . setAppElement ( '#app-base' )
> ```

Resolvel, no meu caso foi Modal.setAppElement("#root"); no component do modal

**Comment by diasbruno** [comment]
To all...

This is the correct behavior for react-modal as already mention on this thread.

As @catamphetamine already mention and many others that have been working with accessibility, there is an old mistake with the default `app element` which uses the `document.body` to add the `aria-hidden` attribute.
This cause a problem because every element that is going to be hidden is a direct descendant of document.body (the modal also becomes inaccessible).

To not break users code, we decided to make `setAppElement` required (or use `appElement` on each case where you use modal), but just warn about the problem to not break stuff...

This also applies to testing. You need to properly setup your tests to use react-modal when needed. 

This is an example to do that (note that you might need to change a little bit depending on the testing framework you use):

```js
let modalNode = null;

before(() => {
  modalNode = document.createElement('div');
  modalNode.id = "modals";
  document.appendChild(modalNode);
  ReactModal.setAppElement('#app');
}):

after(() => {
   document.removeChild(modalNode);
});
```

**Comment by ADTC** [comment]
In Next.js App Router, sometimes the client components could be pre-rendered in the server. I had been using `appElement={document.querySelectorAll('.container')}` but this would fail in the server-side pre-render with `document is not defined`. Adding `?` after `document` will not work.

Using a `try { ... } catch {/* ignore errors */}` with a `let` variable will seem to work, but then I thought a little more deeply about this, and I realized that if the component is pre-rendered **in server**, then it will come to the client with `appElement` set to `undefined`. Then ReactModal will have an `undefined` appElement, possibly for a long time, until the next time the client component is rendered again **in client**.

So I _think_ the below would be the **correct** way to do this. Even if the component is pre-rendered in server, it would hydrate the appElement with the correct element once the pre-render result is transmitted to the client. Is this correct?

```tsx
import { useEffect, useState } from 'react'
import ReactModal from 'react-modal'

// in the component:
  const [appElement, setAppElement] = useState<NodeListOf<Element>>()

  useEffect(() => {
    // When Next.js pre-renders client components in server to generate initial HTML,
    // document is not defined. `?` will not work here, hence the effect & state variable.
    setAppElement(document.querySelectorAll('.container'))
  }, [])

  return (
    <ReactModal
      appElement={appElement}
      // ...
      >
    </ReactModal>
  )
```

_I hope I'm not the first one to figure this out._ 😆

**Comment by romilsondev1** [comment]
In next js 13.5.4 I found the error.

I solved it as follows:

I added the following command to the pages/app.tsx directory

import Modal from 'react-modal';

and before defining my app() function

I added:

Modal.setAppElement('body');
```javascript
import Modal from 'react-modal';

Modal.setAppElement('body');

export default function app(){
```
